File: 25 Chief/javascript/chief.js
// CHIEF
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// Main program
async function main()
{
    print(tab(30) + "CHIEF\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("I AM CHIEF NUMBERS FREEK, THE GREAT INDIAN MATH GOD.\n");
    print("ARE YOU READY TO TAKE THE TEST YOU CALLED ME OUT FOR");
    a = await input();
    if (a.substr(0, 1) != "Y")
        print("SHUT UP, PALE FACE WITH WIE TONGUE.\n");
    print(" TAKE A NUMBER AND ADD 3. DIVIDE THIS NUMBER BY 5 AND\n");
    print("MULTIPLY BY 8. DIVIDE BY 5 AND ADD THE SAME. SUBTRACT 1.\n");
    print("  WHAT DO YOU HAVE");
    b = parseFloat(await input());
    c = (b + 1 - 5) * 5 / 8 * 5 - 3;
    print("I BET YOUR NUMBER WAS " + Math.floor(c + 0.5) + ". AM I RIGHT");
    d = await input();
    if (d.substr(0, 1) != "Y") {
        print("WHAT WAS YOUR ORIGINAL NUMBER");
        k = parseFloat(await input());
        f = k + 3;
        g = f / 5;
        h = g * 8;
        i = h / 5 + 5;
        j = i - 1;
        print("SO YOU THINK YOU'RE SO SMART, EH?\n");
        print("NOW WATCH.\n");
        print(k + " PLUS 3 EQUALS " + f + ". THIS DIVIDED BY 5 EQUALS " + g + ";\n");
        print("THIS TIMES 8 EQUALS " + h + ". IF WE DIVIDE BY 5 AND ADD 5,\n");
        print("WE GET " + i + ", WHICH, MINUS 1, EQUALS " + j + ".\n");
        print("NOW DO YOU BELIEVE ME");
        z = await input();
        if (z.substr(0, 1) != "Y") {
            print("YOU HAVE MADE ME MAD!!!\n");
            print("THERE MUST BE A GREAT LIGHTNING BOLT!\n");
            print("\n");
            print("\n");
            for (x = 30; x >= 22; x--)
                print(tab(x) + "X X\n");
            print(tab(21) + "X XXX\n");
            print(tab(20) + "X   X\n");
            print(tab(19) + "XX X\n");
            for (y = 20; y >= 13; y--)
                print(tab(y) + "X X\n");
            print(tab(12) + "XX\n");
            print(tab(11) + "X\n");
            print(tab(10) + "*\n");
            print("\n");
            print("#########################\n");
            print("\n");
            print("I HOPE YOU BELIEVE ME NOW, FOR YOUR SAKE!!\n");
            return;
        }
    }
    print("BYE!!!\n");
}

main();

File: 25_Chief/lua/chief.lua
--[[
Chief

From: BASIC Computer Games (1978)
Edited by David H. Ahl

    In the words of the program author, John Graham, “CHIEF is designed to
    give people (mostly kids) practice in the four operations (addition,
    multiplication, subtraction, and division).

    It does this while giving people some fun. And then, if the people are
    wrong, it shows them how they should have done it.

    CHIEF was written by John Graham of Upper Brookville, New York.

    
Lua port by Alex Conconi, 2022.
]]--


--- Helper function for tabulating messages.
local function tab(n) return string.rep(" ", n) end


--- Generates a multi-line string representing a lightning bolt
local function bolt()
    local bolt_lines = {}
    for n = 29, 21, -1 do
        table.insert(bolt_lines, tab(n) .. "x x")
    end
    table.insert(bolt_lines, tab(20) .. "x xxx")
    table.insert(bolt_lines, tab(19) .. "x   x")
    table.insert(bolt_lines, tab(18) .. "xx x")
    for n = 19, 12, -1 do
        table.insert(bolt_lines, tab(n) .. "x x")
    end
    table.insert(bolt_lines, tab(11) .. "xx")
    table.insert(bolt_lines, tab(10) .. "x")
    table.insert(bolt_lines, tab(9) .. "*\n")
    table.insert(bolt_lines, string.rep("#", 25) .. "\n")
    return table.concat(bolt_lines, "\n")
end


--- Print the prompt and read a yes/no answer from stdin.
local function ask_yes_or_no(prompt)
    io.stdout:write(prompt .. " ")
    local answer = string.lower(io.stdin:read("*l"))
    -- any line starting with a 'y' or 'Y' is considered a 'yes'
    return answer:sub(1, 1) == "y"
end


--- Print the prompt and read a valid number from stdin.
local function ask_number(prompt)
    io.stdout:write(prompt .. " ")
    while true do
        local n = tonumber(io.stdin:read("*l"))
        if n then
            return n
        else
            print("Enter a valid number.")
        end
    end
end


--- Explain the solution to persuade the player.
local function explain_solution()
    local k = ask_number("What was your original number?")
    -- For clarity we kept the same variable names of the original BASIC version
    local f = k + 3
    local g = f / 5
    local h = g * 8
    local i = h / 5 + 5
    local j = i - 1
    print("So you think you're so smart, eh?")
    print("Now watch.")
    print(k .. " plus 3 equals " .. f .. ". This divided by 5 equals " .. g .. ";")
    print("this times 8 equals " .. h .. ". If we divide by 5 and add 5,")
    print("we get " .. i .. ", which, minus 1, equals " .. j .. ".")
end


--- Main game function.
local function chief_game()
    --- Print game introduction and challenge
    print(tab(29) .. "Chief")
    print(tab(14) .. "Creative Computing  Morristown, New Jersey\n\n")
    print("I am Chief Numbers Freek, the great math god.")
    if not ask_yes_or_no("Are you ready to take the test you called me out for?") then
        print("Shut up, wise tongue.")
    end

    -- Print how to obtain the end result.
    print(" Take a number and add 3. Divide this number by 5 and")
    print("multiply by 8. Divide by 5 and add the same. Subtract 1.")

    -- Ask the result end and reverse calculate the original number.
    local end_result = ask_number(" What do you have?")
    local original_number = (end_result + 1 - 5) * 5 / 8 * 5 - 3

    -- If it is an integer we do not want to print any zero decimals.
    local int_part, dec_part = math.modf(original_number)
    if dec_part == 0 then original_number = int_part end

    -- If the player challenges the answer, print the explanation.
    if not ask_yes_or_no("I bet your number was " .. original_number .. ". Am I right?") then
        explain_solution()
        -- If the player does not accept the explanation, zap them.
        if not ask_yes_or_no("Now do you believe me?") then
            print("YOU HAVE MADE ME MAD!!!")
            print("THERE MUST BE A GREAT LIGHTNING BOLT!\n\n")
            print(bolt())
            print("I hope you believe me now, for your sake!!")
        end
    end
end

--- Run the game.
chief_game()
